File: rest/src/main/java/io/sparkled/viewmodel/playlist/PlaylistViewModelConverterImpl.java
package io.sparkled.viewmodel.playlist;

import io.sparkled.model.entity.Playlist;
import io.sparkled.persistence.playlist.PlaylistPersistenceService;
import io.sparkled.viewmodel.exception.ViewModelConversionException;

import javax.inject.Inject;

public class PlaylistViewModelConverterImpl extends PlaylistViewModelConverter {

    private PlaylistPersistenceService playlistPersistenceService;

    @Inject
    public PlaylistViewModelConverterImpl(PlaylistPersistenceService playlistPersistenceService) {
        this.playlistPersistenceService = playlistPersistenceService;
    }

    @Override
    public PlaylistViewModel toViewModel(Playlist model) {
        return new PlaylistViewModel()
                .setId(model.getId())
                .setName(model.getName());
    }

    @Override
    public Playlist toModel(PlaylistViewModel viewModel) {
        Integer playlistId = viewModel.getId();
        Playlist model = getPlaylist(playlistId);
        return model.setName(viewModel.getName());
    }

    private Playlist getPlaylist(Integer playlistId) {
        if (playlistId == null) {
            return new Playlist();
        }

        return playlistPersistenceService.getPlaylistById(playlistId)
                .orElseThrow(() -> new ViewModelConversionException("Playlist with ID of '" + playlistId + "' not found."));
    }
}

File: rest/src/main/kotlin/io/sparkled/viewmodel/playlist/PlaylistViewModelConverterImpl.kt
package io.sparkled.viewmodel.playlist

import io.sparkled.model.entity.Playlist
import io.sparkled.persistence.playlist.PlaylistPersistenceService
import io.sparkled.viewmodel.exception.ViewModelConversionException

import javax.inject.Inject

class PlaylistViewModelConverterImpl
@Inject constructor(private val playlistPersistenceService: PlaylistPersistenceService) : PlaylistViewModelConverter() {

    override fun toViewModel(model: Playlist): PlaylistViewModel {
        return PlaylistViewModel()
            .setId(model.getId())
            .setName(model.getName())
    }

    override fun toModel(viewModel: PlaylistViewModel): Playlist {
        val playlistId = viewModel.getId()
        val model = getPlaylist(playlistId)

        return model.setName(viewModel.getName())
    }

    private fun getPlaylist(playlistId: Int?): Playlist {
        if (playlistId == null) {
            return Playlist()
        }

        return playlistPersistenceService.getPlaylistById(playlistId)
            ?: throw ViewModelConversionException("Playlist with ID of '$playlistId' not found.")
    }
}
